src/js/Route.js

```diff
@@ -52,23 +54,23 @@ export default class Route {
      * Get whether this route's template matches the given URL.
      *
      * @param {String} url - URL to check.
-     * @return {Boolean} Whether this route matches.
+     * @return {Array|false} If this route matches, returns the matched parameters.
      */
     matchesUrl(url) {
         if (!this.definition.methods.includes('GET')) return false;
 
         // Transform the route's template into a regex that will match a hydrated URL,
         // by replacing its parameter segments with matchers for parameter values
         const pattern = this.template
-            .replace(/\/{[^}?]*\?}/g, '(\/[^/?]+)?')
-            // TODO: the above line with the leading slash is necessary to pick up completely optional *segments*,
-            // like in `/pages/{subPage?}`, so that those are handled first before the more permissive patterns
-            // below, but there's probably a way to do this in one shot
-            .replace(/{[^}?]*\?}/g, '([^/?]+)?')
-            .replace(/{[^}]+}/g, '[^/?]+')
+            .replace(/(\/?){([^}?]*)(\??)}/g, (_, slash, segment, optional) => {
+              const regex = '(?<'+segment+'>' + (this.wheres[segment] || '[^/?]+') + ')';
+              return optional ? '(' + slash + regex + ')?' : slash + regex
+            })
```

Comment:
It would just be pushing the segment name in an array and then match the matched indexes to the segment name, but it does also mean that we need to remove matching groups from the original regex if any (same way laravel does)

Reply:
Okay great. I don't care about IE 11, and Safari 11 usage is tiny so let's keep this!